bluefog/common/operations.cc

```diff
@@ -946,5 +1606,42 @@ bool GetWinOpsWithAssociatedPState() {
   return global_with_associated_p_state;
 }
 
+void SetSkipNegotiateStageState(bool value) {
+  if (!bluefog_global.initialization_done) {
+    BFLOG(ERROR)
+        << "Try to set skip negotiate stage before bluefog is initialized.";
+    return;
+  }
+  if (value == global_skip_negotiate_stage) {
+    return;
+  }
+  if (value) {
+    // From running negotiate to skipping negotiate, we need to properly turn
+    // off negotiate stage. Otherwise, it may hang the processes. Use setting
+    // topology flag to suspend the negotiate stage then skip it.
+    bluefog_global.setting_topology = true;
+    while (!bluefog_global.ready_to_setting_topology.load()) {
+      std::this_thread::sleep_for(std::chrono::microseconds(10));
+    }
+
+    global_skip_negotiate_stage = value;
+
+    bluefog_global.setting_topology = false;
+    bluefog_global.setting_topology_done = true;
+    // Wait for the background thread receive the setting_topology_done and
+    // close the ready_to_setting_topology epoch.
+    while (bluefog_global.ready_to_setting_topology) {
```

Comment:
What is the difference between .load() and not using it?

Reply:
One can specify the memory order. One cannot. It has no big difference here.